README.md
# RTLknitr <img src="man/figures/logo.png" align="right" /> 

[![R-CMD-check](https://github.com/FoadEsmaeili5/RTLknitr/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/FoadEsmaeili5/RTLknitr/actions/workflows/R-CMD-check.yaml)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/RTLknitr)](https://cran.r-project.org/package=RTLknitr)
[![Downloads per month](https://cranlogs.r-pkg.org/badges/RTLknitr)](https://cran.r-project.org/package=RTLknitr)
[![Total Downloads](https://cranlogs.r-pkg.org/badges/grand-total/RTLknitr)](https://cran.r-project.org/package=RTLknitr)
[![Licence](https://img.shields.io/badge/licence-GPL--3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0.en.html)


The R package RTLknitr provide seamless support for right-to-left (RTL) languages, such as Persian and Arabic, in R Markdown documents and LaTeX output. It includes functions and hooks that enable easy integration of RTL language content, allowing users to create documents that adhere to RTL writing conventions.


## Installation

You can install the *stable* version from
CRAN.

```s
install.packages('RTLknitr', dependencies = TRUE)
```

You can install the *development* version from
Github

```s
# install.packages("remotes")
remotes::install_github("FoadEsmaeili5/RTLknitr")
```

## To cite package `RTLknitr` in publications use:
  Esmaeili, F. (2023). _RTLknitr: Right to Left Dynamic Documents Using knitr _. R package version 1.0.0,
  <https://cran.r-project.org/web/packages/RTLknitr>.

A BibTeX entry for LaTeX users is

  @Manual{RTLknitr,
    title = {RTLknitr: Right to Left Dynamic Documents Using knitr},
    author = {Foad Esmaeili},
    year = {2023},
    note = {R package version 1.0.0},
    url = {https://cran.r-project.org/web/packages/RTLknitr}
  }
  


NEWS.md
# CHANGES IN knitr VERSION 1.0.1

Errors in `markdown_rtl_header()` has been fixed.

Errors in the printing of messages in `markdown_rtl()` has been fixed.


R/markdown_rtl.R
#' Enable Right-to-Left (RTL) Formatting for Markdown Documents
#'
#' This function modifies the behavior of \code{\link{knitr}} output hooks to handle RTL (right-to-left)
#' text formatting for Markdown documents, particularly for languages like Persian and Arabic.
#'
#' @details
#' The \code{markdown_rtl} function wraps the existing \code{knit_hooks} with custom commands that switch
#' the text direction between Latin and Persian. It enhances the appearance of the output,
#' messages, warnings, and errors in RTL languages.
#'
#' @param ... Additional arguments to be passed to the underlying \code{knit_hooks$set} function.
#'
#' @return
#' The function has no explicit return value. It modifies the behavior of \code{knit_hooks} for RTL formatting.
#'
#' @export
#'
#' @examples
#' # Enable RTL formatting for the current R Markdown document
#' markdown_rtl()
#'
#' # After enabling RTL formatting, proceed to knit the R Markdown document as usual.
#' # The output and messages will be formatted for RTL languages.
#'
#' @seealso Use \code{\link{hooks_rtl}} for handling RTL formatting within code chunks.
#'
#' @import knitr bookdown
#' @references
#'  Xie, Y. (2014). Dynamic Documents with R and knitr. Chapman & Hall/CRC. ISBN: 978-1482203530.
markdown_rtl <- function(...) {
  .o_o <- knit_hooks$get("output")
  .o_s <- knit_hooks$get("source")
  .o_w <- knit_hooks$get("warning")
  .o_m <- knit_hooks$get("message")
  .o_e <- knit_hooks$get("error")
  .o_i <- knit_hooks$get("inline")
  knit_hooks$set(
    output = function(x, options) {
      paste0(c("\\latin", .o_o(x, options), "\\persian"))
    },
    source = function(x, options) {
      paste0(c("\\latin", .o_s(x, options), "\\persian"))
    },
    warning = function(x, options) {
      paste0(c("\\latin", .o_w(x, options), "\\persian"))
    },
    message = function(x, options) {
      paste0(c("\\latin", .o_m(x, options), "\\persian"))
    },
    error = function(x, options) {
      paste0(c("\\latin", .o_e(x, options), "\\persian"))
    },
    inline = function(x) {
      if (is.numeric(x)) {
        paste0(c("$", .o_i(x), "$"))
      } else {
        .o_i(x)
      }
    }, ...
  )
}


R/tab_header_lbl.R
#' Create a Labeled Header for gt Tables
#'
#' This function enhances the gt package's tab_header function by allowing the inclusion
#' of a label for referencing the table in LaTeX documents.
#'
#' @param data A gt table object.
#' @param title Title of the table.
#' @param subtitle Subtitle of the table.
#' @param label Label for referencing the table in LaTeX.
#'
#' @return LaTeX-formatted table header.
#'
#' @export
#' @details
#' Be sure to use this function at the end of gt format functions. This function
#' returns a LaTeX-formatted table header.
#' @import gt
#' @examples
#' require(gt)
#' head(iris) |> gt() |> tab_header_lbl(title = "iris",label = "tab:iris") |> cat()

tab_header_lbl <- function(data,
                        title,
                        subtitle = NULL,
                        label = NULL) {
  if (is.null(label)) {
    return(gt::tab_header(data, title, subtitle))
  } else {
    gt_tbl <- gt::tab_header(data, title, subtitle) %>%
      gt::as_latex()
    gt_tbl <- gsub(
      "\\caption[*][{]",
      paste("\\caption{\\\\label{", label, "} \\", sep = ""),gt_tbl)
      gsub("labelformat = empty", "labelformat = default", gt_tbl)
  }
}


R/remove_escape_latex.R
#' Remove Escaped LaTeX Formulas from Text
#'
#' This function removes escaped LaTeX formulas and formatting commands from LaTeX-formatted text.
#'
#' @param x A character vector or string containing LaTeX-formatted text.
#'
#' @return A character vector with escaped LaTeX formulas and formatting commands removed.
#'
#' @examples
#' # Remove escaped LaTeX formulas from a string
#' cleaned_text <- remove_escape_latex("\\\\textbf\\{Hello\\}, $\\\\alpha$ formula")
#' print(cleaned_text)
#'
#' @importFrom magrittr %>%
#'
#' @export
remove_escape_latex <- function(x) {
  rm_esc <- gsub(
    "\\\\textbackslash([[:space:]])?",
    "\\\\",
    x,
    fixed = FALSE,
    ignore.case = TRUE
  )
  rm_esc <- gsub("\\\\([&%$#_{}])",
         "\\1",
         rm_esc,
         fixed = FALSE,
         ignore.case = TRUE
    )
    gsub("\\{\\}", "\\\\", rm_esc)
}



R/hooks_rtl.R
#' Customize knitr Hooks for Right-to-Left (RTL) Text Formatting
#'
#' This function defines customized \code{knitr} hooks to handle RTL (right-to-left)
#' text formatting for languages like Persian and Arabic for \code{Sweave}. It allows you to modify
#' the appearance of output, messages, warnings, and errors when producing RTL documents.
#'
#' @param fontcolor Font color for inline output. Default is "black".
#' @param backcolor Background color for inline output. Default is "gray".
#'
#' @return
#' The function returns a list of customized \code{knit_hooks} that handle RTL text formatting.
#' It modifies the output, messages, warnings, errors, inline expressions, and source code
#' to include appropriate commands for RTL languages.
#'
#' @export
#'
#' @examples
#' # Define custom RTL hooks
#' rtl_hooks <- hooks_rtl()
#'
#' # Apply custom RTL hooks to `knit_hooks`
#' knitr::knit_hooks$set(rtl_hooks)
#'
#' # After setting the hooks, knit your document as usual to produce RTL formatting.
#'
#' @seealso Use \code{\link{markdown_rtl}} to enable RTL formatting for Markdown documents.
#'
#' @import knitr
#' @importFrom magrittr %>%
#' @references
#'  Xie, Y. (2014). Dynamic Documents with R and knitr. Chapman & Hall/CRC. ISBN: 978-1482203530.
#'
hooks_rtl <- function(fontcolor = "black", backcolor = "gray") {
  output_old <- knitr::knit_hooks$get("output")
  source_old <- knitr::knit_hooks$get("source")
  message_old <- knitr::knit_hooks$get("message")
  warning_old <- knitr::knit_hooks$get("warning")
  inline_old <- knitr::knit_hooks$get("inline")
  error_old <- knitr::knit_hooks$get("error")

  output_new <- function(x, options) {
    if (options$echo) {
      paste0("\\latin \n", output_old(x, options), "\\persian \n")
    } else {
      output_old(x, options)
    }
  }

  source_new <- function(x, options) {
    if (options$echo) {
      paste0("\\latin \n", source_old(x, options), "\\persian \n")
    } else {
      source_old(x, options)
    }
  }

  message_new <- function(x, options) {
    if (options$echo) {
      paste0("\\latin\n", message_old(x, options), "\\persian\n")
    } else {
      message_old(x, options)
    }
  }

  warning_new <- function(x, options) {
    if (options$echo) {
      paste0("\\latin\n", warning_old(x, options), "\\persian\n")
    } else {
      warning_old(x, options)
    }
  }

  inline_new <- function(x) {
    if (is.numeric(x)) {
      paste0("$", inline_old(x), "$") %>%
        font_back_color(fontcolor = fontcolor, backcolor = backcolor)
    } else {
      inline_old(x) %>%
        font_back_color(fontcolor = fontcolor, backcolor = backcolor)
    }
  }

  error_new <- function(x, options) {
    if (options$echo) {
      paste0("\\latin\n", error_old(x, options), "\\persian\n")
    } else {
      error_old(x, options)
    }
  }

  list(
    error = error_new,
    inline = inline_new,
    warning = warning_new,
    message = message_new,
    source = source_new,
    output = output_new
  )
}
